docs/dev/developer-guides/transactions/fee-model.md

```diff
@@ -39,3 +36,49 @@ So the tips to make the most out of the zkSync fee system are the following:
+
+The code of the validation step of each account can be found in the [DefaultAccount implementation](https://github.com/matter-labs/era-system-contracts/blob/5a6c728576de5db68ad577a09f34e7b85c374192/contracts/DefaultAccount.sol#L65)
+
+
+### Notes on the transaction’s length
+
+zkSync Era sends state diffs onchain, and the cost for the transaction will still mildly depend on its length (because long transactions need to be stored in memory of the operator), also long transactions incur additional costs during interactions with an account. However, the signature (as well as its length) is not available at the time of fee estimation and so there is no correct way to precisely estimate the cost of the transaction. For now, we will compensate for it by multiplying the recommended cost of the transaction by a few percents. In the future, we may introduce the following:
+
+- Each account will be able to implement a method called `fillPartialTransaction` that will fill the signature with the substituted value which will be used for fee estimation.
+
+### Changes for `eth_estimateGas`
+
+Since the entire flow for tx validation + execution will be executed to get the transaction’s fee, the operator will manually put the necessary balance (if needed) for the account before the emulation is run to conduct the default account’s `payForTransaction` method.
+
+
+The `eth_estimateGas` method itself will use binary search to find the smallest gas value under which the transaction *succeeds*.
+
+### Difference from geth
+
+Just like geth, we will use binary search. However, there will be some notable differences in gas estimation from the behavior of geth:
+
+- Unlike geth, it is impossible to track out of gas errors on zkSync Era. The main reason is that the “actual” execution will happen inside the DefaultAccount system contract and due to the 63/64 rule when a high number of gas is provided, the call to the `execute` method of the DefaultAccount will NOT fail due to out of gas even though the subcall to the `transaction.to` contract did fail with an out of gas error.
+
+That means that while geth can distinguish between “empty error” and “out of gas”, we can’t. If during gas estimation we fail to find the necessary `gasLimit`, the following error will be returned _“can not estimate gas: the transaction either fails or requires more gas than allowed”_.
```

Comment:
I checked with the code, the error message is different from the one in the doc, so let's remove this sentence altogether to not confuse people